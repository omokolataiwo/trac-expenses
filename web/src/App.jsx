import React, { Component } from 'react';
import ReactDatePicker from 'react-datepicker';
import 'react-datepicker/dist/react-datepicker.css';
import styled from 'styled-components';
import Expenses from './expenses/Expenses';

const StyledApp = styled.div`
  display: flex;
`;
class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      startDate: new Date(),
      expenses: [],
      expensesForm: {
        description: '',
        amount: '',
        errors: []
      }
    };
    this.handleDateChanged = this.handleDateChanged.bind(this);
    this.handleRegisterExpenses = this.handleRegisterExpenses.bind(this);
    this.changeExpensesFormValues = this.changeExpensesFormValues.bind(this);
  }

  handleDateChanged(startDate) {
    this.setState({ startDate });
  }

  changeExpensesFormValues(event) {
    event.preventDefault();
    const { name, value } = event.target;
    this.setState(prevState => ({
      ...prevState,
      expensesForm: { ...prevState.expensesForm, [name]: value }
    }));
  }

  handleRegisterExpenses() {
    const {
      expensesForm: { description, amount },
      expenses,
      startDate: date
    } = this.state;
    this.setState({
      expenses: [...expenses, { description, amount, date }],
      expensesForm: {
        description: '',
        amount: '',
        errors: []
      }
    });
  }

  render() {
    const { expenses, startDate, expensesForm } = this.state;
    return (
      <StyledApp>
        <div>
          <ReactDatePicker inline startDate={startDate} onChange={this.handleDateChanged} />
        </div>
        <div>
          <Expenses
            expensesForm={expensesForm}
            expenses={expenses}
            handleRegisterExpenses={this.handleRegisterExpenses}
            onExpensesFieldChanged={this.changeExpensesFormValues}
          />
        </div>
      </StyledApp>
    );
  }
}

export default App;
