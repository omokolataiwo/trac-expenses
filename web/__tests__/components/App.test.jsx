import React from 'react';
import { shallow } from 'enzyme';
import App from '../../src/App';

const wrapper = shallow(<App />);
const instance = wrapper.instance();

describe('App Component', () => {
  const descriptionFormValue = 'Transport to office';
  const amountFormValue = '1000';

  it('should render component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('when handleDateChanged is called', () => {
    it('sets the state of startDate', () => {
      const date = 'Tue 12 August 2019';

      instance.handleDateChanged(date);
      expect(wrapper.state().startDate).toEqual(date);
    });
  });

  describe('when changeExpensesFormValues is called', () => {
    let event = { preventDefault: () => {}, target: { description: '', value: '' } };

    it('changes the state of description', () => {
      event = { ...event, target: { name: 'description', value: descriptionFormValue } };
      instance.changeExpensesFormValues(event);
      expect(wrapper.state().expensesForm.description).toEqual(descriptionFormValue);
    });

    it('changes the state of value', () => {
      event = { ...event, target: { name: 'amount', value: amountFormValue } };
      instance.changeExpensesFormValues(event);
      expect(wrapper.state().expensesForm.amount).toEqual(amountFormValue);
    });
  });

  describe('when handleRegisterExpenses is called', () => {
    it('sets the add expenses to state', () => {
      instance.setState({
        expensesForm: { description: descriptionFormValue, amount: amountFormValue }
      });
      instance.handleRegisterExpenses();
      const expectedExpenses = wrapper.state().expenses;
      expect(expectedExpenses.length).toEqual(1);
      expect(expectedExpenses[0].description).toEqual(descriptionFormValue);
    });
  });
});
