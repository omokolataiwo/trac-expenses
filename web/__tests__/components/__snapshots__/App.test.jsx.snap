// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should render component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <div>
          <DatePicker
            allowSameDay={false}
            dateFormat="MM/dd/yyyy"
            dateFormatCalendar="LLLL yyyy"
            disabled={false}
            disabledKeyboardNavigation={false}
            dropdownMode="scroll"
            inline={true}
            inlineFocusSelectedMonth={false}
            monthsShown={1}
            nextMonthButtonLabel="Next month"
            onBlur={[Function]}
            onChange={[Function]}
            onClickOutside={[Function]}
            onFocus={[Function]}
            onInputClick={[Function]}
            onInputError={[Function]}
            onKeyDown={[Function]}
            onMonthChange={[Function]}
            onSelect={[Function]}
            onYearChange={[Function]}
            preventOpenOnFocus={false}
            previousMonthButtonLabel="Previous Month"
            readOnly={false}
            renderDayContents={[Function]}
            shouldCloseOnSelect={true}
            showMonthYearPicker={false}
            showTimeInput={false}
            showTimeSelect={false}
            startDate={2019-08-24T20:59:08.542Z}
            strictParsing={false}
            timeCaption="Time"
            timeInputLabel="Time"
            timeIntervals={30}
            withPortal={false}
          />
        </div>,
        <div>
          <Expenses
            expenses={Array []}
            expensesForm={
              Object {
                "amount": "",
                "description": "",
                "errors": Array [],
              }
            }
            handleRegisterExpenses={[Function]}
            onExpensesFieldChanged={[Function]}
          />
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <DatePicker
            allowSameDay={false}
            dateFormat="MM/dd/yyyy"
            dateFormatCalendar="LLLL yyyy"
            disabled={false}
            disabledKeyboardNavigation={false}
            dropdownMode="scroll"
            inline={true}
            inlineFocusSelectedMonth={false}
            monthsShown={1}
            nextMonthButtonLabel="Next month"
            onBlur={[Function]}
            onChange={[Function]}
            onClickOutside={[Function]}
            onFocus={[Function]}
            onInputClick={[Function]}
            onInputError={[Function]}
            onKeyDown={[Function]}
            onMonthChange={[Function]}
            onSelect={[Function]}
            onYearChange={[Function]}
            preventOpenOnFocus={false}
            previousMonthButtonLabel="Previous Month"
            readOnly={false}
            renderDayContents={[Function]}
            shouldCloseOnSelect={true}
            showMonthYearPicker={false}
            showTimeInput={false}
            showTimeSelect={false}
            startDate={2019-08-24T20:59:08.542Z}
            strictParsing={false}
            timeCaption="Time"
            timeInputLabel="Time"
            timeIntervals={30}
            withPortal={false}
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "allowSameDay": false,
            "dateFormat": "MM/dd/yyyy",
            "dateFormatCalendar": "LLLL yyyy",
            "disabled": false,
            "disabledKeyboardNavigation": false,
            "dropdownMode": "scroll",
            "inline": true,
            "inlineFocusSelectedMonth": false,
            "monthsShown": 1,
            "nextMonthButtonLabel": "Next month",
            "onBlur": [Function],
            "onChange": [Function],
            "onClickOutside": [Function],
            "onFocus": [Function],
            "onInputClick": [Function],
            "onInputError": [Function],
            "onKeyDown": [Function],
            "onMonthChange": [Function],
            "onSelect": [Function],
            "onYearChange": [Function],
            "preventOpenOnFocus": false,
            "previousMonthButtonLabel": "Previous Month",
            "readOnly": false,
            "renderDayContents": [Function],
            "shouldCloseOnSelect": true,
            "showMonthYearPicker": false,
            "showTimeInput": false,
            "showTimeSelect": false,
            "startDate": 2019-08-24T20:59:08.542Z,
            "strictParsing": false,
            "timeCaption": "Time",
            "timeInputLabel": "Time",
            "timeIntervals": 30,
            "withPortal": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Expenses
            expenses={Array []}
            expensesForm={
              Object {
                "amount": "",
                "description": "",
                "errors": Array [],
              }
            }
            handleRegisterExpenses={[Function]}
            onExpensesFieldChanged={[Function]}
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "expenses": Array [],
            "expensesForm": Object {
              "amount": "",
              "description": "",
              "errors": Array [],
            },
            "handleRegisterExpenses": [Function],
            "onExpensesFieldChanged": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-bwzfXH",
        "isStatic": true,
        "rules": Array [
          "
  display: flex;
",
        ],
      },
      "displayName": "styled.div",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-bwzfXH",
      "target": "div",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <div>
            <DatePicker
              allowSameDay={false}
              dateFormat="MM/dd/yyyy"
              dateFormatCalendar="LLLL yyyy"
              disabled={false}
              disabledKeyboardNavigation={false}
              dropdownMode="scroll"
              inline={true}
              inlineFocusSelectedMonth={false}
              monthsShown={1}
              nextMonthButtonLabel="Next month"
              onBlur={[Function]}
              onChange={[Function]}
              onClickOutside={[Function]}
              onFocus={[Function]}
              onInputClick={[Function]}
              onInputError={[Function]}
              onKeyDown={[Function]}
              onMonthChange={[Function]}
              onSelect={[Function]}
              onYearChange={[Function]}
              preventOpenOnFocus={false}
              previousMonthButtonLabel="Previous Month"
              readOnly={false}
              renderDayContents={[Function]}
              shouldCloseOnSelect={true}
              showMonthYearPicker={false}
              showTimeInput={false}
              showTimeSelect={false}
              startDate={2019-08-24T20:59:08.542Z}
              strictParsing={false}
              timeCaption="Time"
              timeInputLabel="Time"
              timeIntervals={30}
              withPortal={false}
            />
          </div>,
          <div>
            <Expenses
              expenses={Array []}
              expensesForm={
                Object {
                  "amount": "",
                  "description": "",
                  "errors": Array [],
                }
              }
              handleRegisterExpenses={[Function]}
              onExpensesFieldChanged={[Function]}
            />
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <DatePicker
              allowSameDay={false}
              dateFormat="MM/dd/yyyy"
              dateFormatCalendar="LLLL yyyy"
              disabled={false}
              disabledKeyboardNavigation={false}
              dropdownMode="scroll"
              inline={true}
              inlineFocusSelectedMonth={false}
              monthsShown={1}
              nextMonthButtonLabel="Next month"
              onBlur={[Function]}
              onChange={[Function]}
              onClickOutside={[Function]}
              onFocus={[Function]}
              onInputClick={[Function]}
              onInputError={[Function]}
              onKeyDown={[Function]}
              onMonthChange={[Function]}
              onSelect={[Function]}
              onYearChange={[Function]}
              preventOpenOnFocus={false}
              previousMonthButtonLabel="Previous Month"
              readOnly={false}
              renderDayContents={[Function]}
              shouldCloseOnSelect={true}
              showMonthYearPicker={false}
              showTimeInput={false}
              showTimeSelect={false}
              startDate={2019-08-24T20:59:08.542Z}
              strictParsing={false}
              timeCaption="Time"
              timeInputLabel="Time"
              timeIntervals={30}
              withPortal={false}
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "allowSameDay": false,
              "dateFormat": "MM/dd/yyyy",
              "dateFormatCalendar": "LLLL yyyy",
              "disabled": false,
              "disabledKeyboardNavigation": false,
              "dropdownMode": "scroll",
              "inline": true,
              "inlineFocusSelectedMonth": false,
              "monthsShown": 1,
              "nextMonthButtonLabel": "Next month",
              "onBlur": [Function],
              "onChange": [Function],
              "onClickOutside": [Function],
              "onFocus": [Function],
              "onInputClick": [Function],
              "onInputError": [Function],
              "onKeyDown": [Function],
              "onMonthChange": [Function],
              "onSelect": [Function],
              "onYearChange": [Function],
              "preventOpenOnFocus": false,
              "previousMonthButtonLabel": "Previous Month",
              "readOnly": false,
              "renderDayContents": [Function],
              "shouldCloseOnSelect": true,
              "showMonthYearPicker": false,
              "showTimeInput": false,
              "showTimeSelect": false,
              "startDate": 2019-08-24T20:59:08.542Z,
              "strictParsing": false,
              "timeCaption": "Time",
              "timeInputLabel": "Time",
              "timeIntervals": 30,
              "withPortal": false,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Expenses
              expenses={Array []}
              expensesForm={
                Object {
                  "amount": "",
                  "description": "",
                  "errors": Array [],
                }
              }
              handleRegisterExpenses={[Function]}
              onExpensesFieldChanged={[Function]}
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "expenses": Array [],
              "expensesForm": Object {
                "amount": "",
                "description": "",
                "errors": Array [],
              },
              "handleRegisterExpenses": [Function],
              "onExpensesFieldChanged": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-bwzfXH",
          "isStatic": true,
          "rules": Array [
            "
  display: flex;
",
          ],
        },
        "displayName": "styled.div",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-bwzfXH",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
